File: 22. Generate Parentheses.cpp
class Solution {
public:
    // backtracking : 83.75%, ok
    vector<string> generateParenthesis(int n) {
        vector<string> ret;
        if(!n) return vector<string>(1, "");
        helper("", n, n, ret);
        return ret;
    }

    void helper(string curr, int left, int right, vector<string>& ret){
        if(!left && !right){
            ret.push_back(curr);
            return;
        }
        if(left) helper(curr + "(", left - 1, right, ret);
        if(right > left)
            helper(curr + ")", left, right - 1, ret);
    }
};

File: 22. Generate Parentheses.py
class Solution(object):
    
#     # brute force: O(2^N)time ??? & space, TLE
#     def generateParenthesis(self, n):
#         """
#         :type n: int
#         :rtype: List[str]
#         """
#         if n == 0: return []
#         self.ret, self.n = set(), n
#         self.helper(n, n, '')
#         return list(self.ret)
    
#     def helper(self, left, right, curr):
#         if left == right == 0:
#             self.ret.add(curr)
#         else:
#             for i in xrange(1, left+1):
#                 self.helper(left-i, right, curr + '('*i)
#             for j in xrange(1, right-left+1):
#                 self.helper(left, right-j, curr + ')'*j)


    # back tracking: O(2^N)time ??? & space, TLE
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        if n == 0: return []
        self.ret, self.n = [], n
        self.helper(n, n, '')
        return self.ret
    
    def helper(self, left, right, curr):
        if left == right == 0:
            self.ret.append(curr)
        else:
            if left:
                self.helper(left-1, right, curr + '(')
            if right > left:
                self.helper(left, right-1, curr + ')')